OrbitVulkanLayer/SubmissionTracker.h

```diff
@@ -0,0 +1,779 @@
+          if (marker.slot_index.has_value() && queue_submission_optional.has_value()) {
+            submitted_marker = {.meta_information = queue_submission_optional->meta_information,
+                                .slot_index = marker.slot_index.value()};
+          }
+
+          switch (marker.type) {
+            case MarkerType::kDebugMarkerBegin: {
+              if (queue_submission_optional.has_value() && marker.slot_index.has_value()) {
+                ++queue_submission_optional->num_begin_markers;
+              }
+              CHECK(marker.text.has_value());
+              CHECK(marker.color.has_value());
+              internal::MarkerState marker_state{.label_name = marker.text.value(),
+                                                 .color = marker.color.value(),
+                                                 .begin_info = submitted_marker,
+                                                 .depth = markers.marker_stack.size(),
+                                                 .cut_off = marker.cut_off};
+              markers.marker_stack.push(std::move(marker_state));
+              break;
+            }
+
+            case MarkerType::kDebugMarkerEnd: {
+              internal::MarkerState marker_state = markers.marker_stack.top();
+              markers.marker_stack.pop();
+
+              // If there is a begin marker slot from a previous submission, this is our chance to
+              // reset the slot.
+              if (marker_state.begin_info.has_value() && !queue_submission_optional.has_value()) {
+                marker_slots_to_reset.push_back(marker_state.begin_info.value().slot_index);
+              }
+
+              // If the begin marker was cut off, but the end marker was not (as it is in a
+              // different submission), we reset the slot
+              if (marker_state.cut_off && marker.slot_index.has_value()) {
+                marker_slots_to_reset.push_back(marker.slot_index.value());
+              }
+
+              if (queue_submission_optional.has_value() && marker.slot_index.has_value() &&
+                  !marker_state.cut_off) {
+                marker_state.end_info = submitted_marker;
+                queue_submission_optional->completed_markers.emplace_back(std::move(marker_state));
+              }
+
+              break;
+            }
+          }
+        }
+        command_buffer_to_state_.erase(command_buffer);
+      }
+    }
+
+    if (!marker_slots_to_reset.empty()) {
+      timer_query_pool_->ResetQuerySlots(device, marker_slots_to_reset);
+    }
+
+    if (!queue_submission_optional.has_value()) {
+      return;
+    }
+
+    if (!queue_to_submissions_.contains(queue)) {
+      queue_to_submissions_[queue] = {};
+    }
+    CHECK(queue_to_submissions_.contains(queue));
+    queue_to_submissions_.at(queue).emplace_back(std::move(queue_submission_optional.value()));
+  }
+
+  void CompleteSubmits(VkDevice device) {
```

Comment:
I would also add an explanatory comment to this method as it's quite an important one.

Reply:
Totally, agree. Completely missed that method. Added a comment. Let me know what you think.